### PR TITLE
Document the core::upgrade module

### PR DESCRIPTION
**Pull request by tomaka**
Adds doc-comments to the `core::upgrade` module.

### CODE REVIEW

File: core/src/upgrade/map.rs [hunk 1]
```diff
@@ -21,6 +21,7 @@
 use futures::prelude::*;
 use crate::upgrade::{InboundUpgrade, OutboundUpgrade, UpgradeInfo};
 
+/// Wraps around an upgrade and applies a closure on the output.
```
Comment (dvdplm) [inline core/src/upgrade/map.rs]:
```suggestion
/// Wraps around an upgrade and applies a closure to the output.
```

File: core/src/upgrade/map.rs [hunk 2]
```diff
@@ -76,6 +77,7 @@ where
     }
 }
 
+/// Wraps around an upgrade and applies a closure on the error.
```
Comment (dvdplm) [inline core/src/upgrade/map.rs]:
```suggestion
/// Wraps around an upgrade and applies a closure to the error.
```

File: core/src/upgrade/mod.rs [hunk 3]
```diff
@@ -18,6 +18,45 @@
 // FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 // DEALINGS IN THE SOFTWARE.
 
+//! Contains everything related to upgrading a connection or a substream to use a protocol.
+//!
+//! After a connection with a remote has been successfully established or a substream successfully
+//! opened, the next step is to *upgrade* this connection or substream to use a protocol.
+//!
+//! This is where the `UpgradeInfo`, `InboundUpgrade` and `OutboundUpgrade` traits come into play.
+//! The `InboundUpgrade` and `OutboundUpgrade` traits are implemented on types that represent an
```
Comment (dvdplm) [inline core/src/upgrade/mod.rs]:
```suggestion
//! The `InboundUpgrade` and `OutboundUpgrade` traits are implemented on types that represent a
```

File: core/src/upgrade/mod.rs [hunk 4]
```diff
@@ -18,6 +18,45 @@
...
+//! collection of one or more possible protocols for respectively an ingoing or outgoing
+//! connection or substream.
+//!
+//! > **Note**: Multiple versions of the same protocol are handled the same way as multiple
```
Comment (dvdplm) [inline core/src/upgrade/mod.rs]: Maybe:
```suggestion
//! > **Note**: Multiple versions of the same protocol are treated as different protocols. For example `/foo/1.0.0` and `/foo/1.1.0` are totally unrelated as far as upgrading is concerned.
```

File: core/src/upgrade/mod.rs [hunk 5]
```diff
@@ -18,6 +18,45 @@
...
+//!
+//! - A protocol negotiation step. The `UpgradeInfo::protocol_names` method is called to determine
+//!   which protocols are supported by the trait implementation. The `multistream-select` protocol
+//!   is used in order to agree on which protocol to use amongst the ones supported.
+//!
+//! - A handshake. After a successful negotiation, the `InboundUpgrade::upgrade` or
+//!   `OutboundUpgrade::upgrade` method is called. This method will return a `Future` that
+//!   performs a handshake. This handshake is considered mandatory, however in practice it is
+//!   possible for the trait implementation to return a dummy `Future` that doesn't perform any
+//!   action and immediately succeeds.
```
Comment (dvdplm) [inline core/src/upgrade/mod.rs]: q: what would a use-case for "skipping" the handshake like this be?

File: core/src/upgrade/mod.rs [hunk 6]
```diff
@@ -37,22 +76,40 @@ pub use self::{
...
     type Future: Future<Item = Self::Output, Error = Self::Error>;
 
+    /// After we have determined that the remote supports one of the protocols we support, this
+    /// method is called to start the handshake.
+    ///
+    /// The `id` is the identifier of the protocol, as produced by `protocol_names()`.
     fn upgrade_inbound(self, socket: C, id: Self::UpgradeId) -> Self::Future;
 }
 
+/// Extention trait for `InboundUpgrade`. Automatically implemented on all types that implement
```
Comment (dvdplm) [inline core/src/upgrade/mod.rs]:
```suggestion
/// Extension trait for `InboundUpgrade`. Automatically implemented on all types that implement
```

File: core/src/upgrade/mod.rs [hunk 7]
```diff
@@ -37,22 +76,40 @@ pub use self::{
...
+/// `InboundUpgrade`.
 pub trait InboundUpgradeExt<C>: InboundUpgrade<C> {
+    /// Returns a new object that wraps around `Self` and applies a closure on the `Output`.
```
Comment (dvdplm) [inline core/src/upgrade/mod.rs]:
```suggestion
    /// Returns a new object that wraps around `Self` and applies a closure to the `Output`.
```

File: core/src/upgrade/mod.rs [hunk 8]
```diff
@@ -61,6 +118,7 @@ pub trait InboundUpgradeExt<C>: InboundUpgrade<C> {
         MapUpgrade::new(self, f)
     }
 
+    /// Returns a new object that wraps around `Self` and applies a closure on the `Error`.
```
Comment (dvdplm) [inline core/src/upgrade/mod.rs]:
```suggestion
    /// Returns a new object that wraps around `Self` and applies a closure to the `Error`.
```

File: core/src/upgrade/mod.rs [hunk 9]
```diff
@@ -80,15 +140,26 @@ pub trait InboundUpgradeExt<C>: InboundUpgrade<C> {
...
+    /// method is called to start the handshake.
+    ///
+    /// The `id` is the identifier of the protocol, as produced by `protocol_names()`.
     fn upgrade_outbound(self, socket: C, id: Self::UpgradeId) -> Self::Future;
 }
 
+/// Extention trait for `OutboundUpgrade`. Automatically implemented on all types that implement
+/// `OutboundUpgrade`.
 pub trait OutboundUpgradeExt<C>: OutboundUpgrade<C> {
+    /// Returns a new object that wraps around `Self` and applies a closure on the `Output`.
```
Comment (dvdplm) [inline core/src/upgrade/mod.rs]:
```suggestion
    /// Returns a new object that wraps around `Self` and applies a closure to the `Output`.
```

File: core/src/upgrade/mod.rs [hunk 10]
```diff
@@ -97,6 +168,7 @@ pub trait OutboundUpgradeExt<C>: OutboundUpgrade<C> {
         MapUpgrade::new(self, f)
     }
 
+    /// Returns a new object that wraps around `Self` and applies a closure on the `Error`.
```
Comment (dvdplm) [inline core/src/upgrade/mod.rs]:
```suggestion
    /// Returns a new object that wraps around `Self` and applies a closure to the `Error`.
```

File: core/src/upgrade/mod.rs (hunk 5 above)
Comment (tomaka) [reply core/src/upgrade/mod.rs]: Almost all protocols skip the handshake. The exceptions are secio and identify.
What I call the handshake is the step between the negotiation and the output.

**Pull request merged**